pkg/pki/ca/ca.go

```diff
@@ -80,6 +105,47 @@ func NewSelfSignedIstioCA(caCertTTL, certTTL time.Duration, org string) (*IstioC
 // NewIstioCA returns a new IstioCA instance.
 func NewIstioCA(opts *IstioCAOptions) (*IstioCA, error) {
 	ca := &IstioCA{certTTL: opts.CertTTL}
+	// If the signing key/cert or root cert is empty, we should create a self-signed key/cert pair,
+	// and write it to the secret for persistent purpose.
+	// TODO(wattli): get rid of the NewSelfSignedIstioCA() after 0.2.
+	// K8s secret has a bug that you can't specify empty content, so we put a small content in it and
+	// use 10 as a threshold.
+	if len(opts.RootCertBytes) < 10 || len(opts.SigningCertBytes) < 10 || len(opts.SigningKeyBytes) < 10 {
+		now := time.Now()
+		options := CertOptions{
+			NotBefore:    now,
+			NotAfter:     now.Add(opts.CertTTL),
+			Org:          opts.Org,
+			IsCA:         true,
+			IsSelfSigned: true,
+			RSAKeySize:   caKeySize,
+		}
+		pemCert, pemKey := GenCert(options)
+		opts.CertChainBytes = []byte{}
+		opts.RootCertBytes = pemCert
+		opts.SigningCertBytes = pemCert
+		opts.SigningKeyBytes = pemKey
+
+		// Rewrite the key/cert back to secret so they will be persistent when CA restarts.
+		secret := &apiv1.Secret{
+			Data: map[string][]byte{
+				CACertChainID:  pemCert,
+				CAPrivateKeyID: pemKey,
+				RootCertID:     pemCert,
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Annotations: map[string]string{serviceAccountNameAnnotationKey: CAServiceAccount},
+				Name:        SecretNamePrefix + CAServiceAccount,
```

Comment:
So if an admin specifies an empty secret called `ca-signing-secret` and mounts it into the CA pod, the CA will in turn write the self-signed key/cert into a totally different secret (e.g. `istio-ca.istio-ca-creds`). This would be quite confusing, right? And it won't help in case the ca pod is recreated.

Reply:
Good point, an easy way is to document it clearly not to mess up with the default CA config. It might worth to make it a flag going forward.

Reply:
Good point. The secret (with dummy content) needs to be there before deploying istio-ca, and we will have a yaml file to create that secret (in front of istio-ca yaml). 

An easy way is to document it clearly not to mess up the default CA secret name (and I doubt people will actually do that). 

Of course this can be done via flag, but we will need admin to specify secret name at two places: where we do volume mount and CA input params. Personally I prefer not to introduce another flag to CA.

Reply:
hmm I feel it's quite a restriction to force the secret to be named as `isito-ca.istio-ca-creds`. 

IMO the idea workflow would be:

1. Admin creates a secret with an arbitrarily chosen name
1. The admin mounts the secret into CA at a well-defined dir
1. CA examines the dir, and write key/cert if the dir is empty
1. The content that CA writes is auto fed back to apiserver (we might need to verify if this is possible)

If step 4 is not possible, I'd vote to explicitly passing in a secret name.